functions/send_encounters/main.py

```diff
@@ -0,0 +1,168 @@
+            }), 401
+
+    upload_id = secrets.token_hex(32)
+
+    _save_encounter_uploads_to_datastore(user_id, upload_id)
+    if not _save_encounters_to_bigquery(user_id, upload_id, encounters):
+        return jsonify(
+            {
+                'status': 'failed',
+                'message': 'Internal error'
+            }), 500
+
+    _update_user_entity(user_entity, platform, os_version, app_version, device_type, lang)
+    return jsonify({"status": "OK"})
+
+
+def _parse_request(request: Request) -> dict:
+    if not request.is_json:
+        raise InvalidRequestException(
+            422,
+            {
+                'status': 'failed',
+                'message': 'invalid data'
+            }
+        )
+
+    request_data = request.get_json()
+
+    for key in [KEY_USER_ID, KEY_PLATFORM, KEY_OS_VERSION, KEY_DEVICE_TYPE, KEY_APP_VERSION, KEY_LANG, KEY_ENCOUNTERS]:
+        if key not in request_data:
+            raise InvalidRequestException(
+                422,
+                {
+                    'status': 'failed',
+                    'message': f'missing field: {key}'
+                }
+            )
+
+    for encounter in request_data["encounters"]:
+        for key in [KEY_ENCOUNTER_DATE, KEY_BEACON_ID, KEY_SIGNAL_STRENGTH]:
+            if key not in encounter:
+                raise InvalidRequestException(
+                    422,
+                    {
+                        'status': 'failed',
+                        'message': f'missing field in encounters: {key}'
+                    }
+                )
+
+    return request_data
+
+
+def _get_user_entity(user_id: str) -> Optional[Entity]:
+    device_key = datastore_client.key(USERS_DATASTORE_KIND, f"{user_id}")
+    return datastore_client.get(key=device_key)
+
+
+def _save_encounter_uploads_to_datastore(user_id: str, upload_id: str) -> None:
+    uploads_key = datastore_client.key(ENCOUNTER_UPLOADS_DATASTORE_KIND, f"{upload_id}")
+    entity = datastore.Entity(key=uploads_key)
+
+    entity.update({
+        'user_id': user_id,
+        'upload_id': upload_id,
+        'date': datetime.utcnow(),
+        'processed_by_health_authority': False,
+        'confirmed_by_health_authority': None,
+    })
+    datastore_client.put(entity)
+
+
+def _save_encounters_to_bigquery(user_id: str, upload_id: str, encounters: list) -> bool:
+    bq_client = bigquery.Client()
+    table = bq_client.get_table(BQ_TABLE_ID)
+
+    rows_to_insert = []
+    for encounter in encounters:
+        encounter_date = datetime.strptime(encounter[KEY_ENCOUNTER_DATE], "%Y%m%d%H")
+        beacon_id = encounter[KEY_BEACON_ID]
+        signal_strength = encounter[KEY_SIGNAL_STRENGTH]
+        rows_to_insert.append(
+            (
+                upload_id,
+                user_id,
+                encounter_date,
+                beacon_id,
+                signal_strength
+            )
+        )
+
+    errors = bq_client.insert_rows(table, rows_to_insert)
```

Comment:
Na stronie [Quotas and Limits](https://cloud.google.com/bigquery/quotas#streaming_inserts) można znaleźć informację o maksymalnym rozmiarze requestu HTTP w przypadku Streaming inserts (używane przez `insert_rows`):
> HTTP request size limit: 10 MB (see Note)
> Exceeding this value will cause invalid errors.
> Note: Internally the request is translated from HTTP json into an internal data structure. The translated data structure has its own enforced size limit. It's hard to predict the size of the resulting internal data structure, but the chance of hitting the internal limit is very low if you keep your HTTP requests to 5 MB or less.

Oczywiście 5 MB to sporo i jest szansa że to nigdy się nie wydarzy (zależy jak działa aplikacja mobilna) ale myślę, że z uwagi na fakt, że te dane są niezwykle ważne i nie wiadomo kiedy i czy nastąpi retry warto rozważyć trzymanie się limitu 5 MB przy dodawaniu tych danych. W praktyce trzeba wywołać `insert_rows` kilka razy z mniejszymi paczkami.

Reply:
Dzięki za info - czy mógłbyś proszę przekuć to w Issue? Zajmiemy się tym wkrótce.